### DISCUSSION TITLE
Large db with millions of rows [Nov 12]

### DISCUSSION THREAD

**Discussion by dstala** [nocodb/nocodb#4440] [Discord Threads]
> 🙋‍♂️ **drasticstuff**

**Comment by o1lab** [comment]
> 💁 **o1lab**  
#### explore should be fine   
#### if filter is applied on unoptimised columns like text or to columns which doesnt have an index - you can expect to be slow at db level  
#  
> 🙋‍♂️ **drasticstuff**  
#### Thanks, so just to be clear there is no problem at the NocoDB front end? I assume the rows are paginated?  
#  
> 🙋‍♂️ **drasticstuff**  
#### Doesn't seem to be working, I get death spinner when selecting the table to view it  
#  
> 🙋‍♂️ **drasticstuff**  
#### This is with SQLite and table with ~100 million rows  
#  
> 🙋‍♂️ **Braks**  
#### Have you received any errors in the console? Or just a blank spinner page (the page seems to be loading for a *long* time)?  
#  
> 💁 **o1lab**  
#### how large is the sqlitefile ?  
#  
> 🙋‍♂️ **drasticstuff**  
#### 21 GB  
#  
> 🙋‍♂️ **drasticstuff**  
#### I don't see anything in the console  
#  
> 🙋‍♂️ **drasticstuff**  
#### Just the spinner (the entire page is not blank, only the part where the table would show up)  
#  
> 🙋‍♂️ **drasticstuff**  
#### Okay, it did finally load. It took a few minutes. Is that expected? Looking at the output of `top` it seems node is CPU-bound (but I would have expected something I/O-bound)  
#  
> 💁 **o1lab**  
#### was the noco process cpu bound ? how much RAM do you have .. 21gb is a large file.. so sqlite would need good enough memory  
#  
> 🙋‍♂️ **drasticstuff**  
#### I have 32GB ram, I have no problem using the file using the sqlite3 CLI or programmatically. I didn’t see a noco process in `atop`  
#  
> 🙋‍♂️ **drasticstuff**  
#### Same problem with postgres  
#  
> 🙋‍♂️ **drasticstuff**  
#### (Though now it shows postgres as disk-bound)  
#  
> 🙋‍♂️ **drasticstuff**  
#### I'm guessing it's doing a full table scan for some reason.  
#  
> 🙋‍♂️ **drasticstuff**  
#### Unfortunate  
#  
> 💁 **o1lab**  
#### does table have a PK, is it auto increment and does it have created_at   
#### we usually sort it by primary key or created_at   
#### if both arent there - its a simple select * with limit  
#  
> 💁 **o1lab**  
#### you can also try querying from our apis  
#  
> 💁 **o1lab**  
#### to see whether there is any difference  
#  
> 🙋‍♂️ **drasticstuff**  
#### No PK but if there is a sort that would definitely be causing a problem.  I don't know much about postgres but I'll try to see if I can get it to log the queries  
#  
> 🙋‍♂️ **drasticstuff**  
#### So I turned on `auto_explain`. The problem is that nocodb is running `select count(*)` which is a problem for big databases.  
#  
> 🙋‍♂️ **drasticstuff**  
#### Maybe in `nc_configureFields`  
#  
> 🙋‍♂️ **drasticstuff**  
#### (job.ts)  
#  
> 💁 **o1lab**  
#### Ah thanks  
#  
> 🙋‍♂️ **drasticstuff**  
#### Is it possible that this behavior will be changed?  
#  
> 💁 **o1lab**  
#### we can improve count(*) to count(pk) but having a table which doesnt have PK makes me wonder how best we can go for it  
#  
> 💁 **o1lab**  
#### will file a github issue - lets see  
#  
> 💁 **o1lab**  
#### if you can send a PR that will be helpful  
#  
> 💁 **o1lab**  
#### https://github.com/nocodb/nocodb/issues/4287  
#  
> 🙋‍♂️ **drasticstuff**  
#### I can add a PK but that doesn’t help avoid a full scan does it?  
#  
> 💁 **o1lab**  
#### if there is a pk, there is an index and count can take less time is my opinion  
#  
> 🙋‍♂️ **drasticstuff**  
#### The problem is that it has to go through every item; there is no way to get a fast count on an SQL database as far as I know. I don’t know if it’s any  faster with an index but it’s not going to be nearly fast enough to make large databases useable.  
#  
> 🙋‍♂️ **drasticstuff**  
#### That’s okay not every product supports every use case  
#  

 [Autogenerated from discord](https://discord.com/channels/661905455894888490/1037400087439081532)  
- - - -  
#### Join [NocoDB's](http://www.nocodb.com/)  community  
[Discord](https://discord.gg/5RgZmkW) • [Twitter](https://twitter.com/nocodb) • [Reddit](https://www.reddit.com/r/NocoDB/) • [Docs](https://docs.nocodb.com/)
